code/modules/telesci/gps.dm

```diff
@@ -13,69 +11,145 @@ var/global/list/gps_by_type = list()
 	var/gps_prefix = "COM"
 	var/gpstag = "COM0"
 	var/emped = 0
+	var/held_by = null
 	var/turf/locked_location
+	var/list/tracking = list()
 
 /obj/item/device/gps/Initialize()
 	. = ..()
-	GPS_list += src
-	LAZYADD(gps_by_type["[type]"], src)
-	gpstag = "[gps_prefix][LAZYLEN(gps_by_type["[type]"])]"
+	gpstag = get_initial_tag()
 	name = "global positioning system ([gpstag])"
+	update_position()
 	add_overlay("working")
+	moved_event.register(src, src, /obj/item/device/gps/proc/update_position)
+
+	for(var/gps in GPS_list)
+		tracking += GPS_list[gps]["tag"]
 
 /obj/item/device/gps/Destroy()
-	GPS_list -= src
-	var/list/typelist = gps_by_type["[type]"]
-	LAZYREMOVE(typelist, src)
+	GPS_list -= GPS_list[gpstag]
+	moved_event.unregister(src, src)
+	if(held_by)
+		moved_event.unregister(held_by, src)
+		held_by = null
 	return ..()
 
+/obj/item/device/gps/pickup(var/mob/user)
+	..()
+	held_by = user
+	moved_event.register(user, src, /obj/item/device/gps/proc/update_position)
+
+/obj/item/device/gps/dropped(var/mob/user)
+	..()
+	held_by = null
+	moved_event.unregister(user, src)
+
 /obj/item/device/gps/emp_act(severity)
 	emped = 1
 	cut_overlay("working")
 	add_overlay("emp")
 	addtimer(CALLBACK(src, .proc/post_emp), 300)
+	update_position()
 
 /obj/item/device/gps/proc/post_emp()
 	emped = 0
 	cut_overlay("emp")
 	add_overlay("working")
+	update_position()
 
-/obj/item/device/gps/attack_self(mob/user)
+/obj/item/device/gps/vueui_data_change(var/list/data, var/mob/user, var/datum/vueui/ui)
+	if(!data)
+		. = data = list()
+	VUEUI_SET_CHECK(data["own_tag"], gpstag, ., data)
+	
+	var/list/tracking_list = list()
+	for(var/tracking_tag in sortList(tracking))
+		if(!GPS_list[tracking_tag]) // Another GPS device has changed its tag or been destroyed
+			tracking -= tracking_tag
+			continue
 
-	var/obj/item/device/gps/t = ""
-	var/gps_window_height = 110 + GPS_list.len * 20 // Variable window height, depending on how many GPS units there are to show
-	if(emped)
-		t += "ERROR"
-	else
-		t += "<BR><A href='?src=\ref[src];tag=1'>Set Tag</A> "
-		t += "<BR>Tag: [gpstag]"
-		if(locked_location && locked_location.loc)
-			t += "<BR>Bluespace coordinates saved: [locked_location.loc]"
-			gps_window_height += 20
-
-		for(var/obj/item/device/gps/G in GPS_list)
-			var/turf/pos = get_turf(G)
-			var/area/gps_area = get_area(G)
-			var/tracked_gpstag = G.gpstag
-			if(G.emped == 1 || !pos)
-				t += "<BR>[tracked_gpstag]: ERROR"
-			else
-				t += "<BR>[tracked_gpstag]: [format_text(gps_area.name)] ([pos.x], [pos.y], [pos.z])"
+		if(AreConnectedZLevels(src.loc.z, GPS_list[tracking_tag]["pos_z"]))
+			tracking_list[tracking_tag] = (GPS_list[tracking_tag])
+
+	data["tracking_list"] = tracking_list
```

Comment:
It won't update correctly, because this function returnsd only when tag is changed. As it being properly checked. At https://github.com/Aurorastation/Aurora.3/pull/8076/files#diff-aa0af895862ff9fa692a0d05303a2d3eR63

Reply:
It was updating properly in testing. Hmmm.

Reply:
Changed it to use the proper watchy var and set var stuff.